internal/hcsoci/assigned_devices.go

```diff
@@ -0,0 +1,157 @@
+		return nil, fmt.Errorf("no assignable devices on the spec %v", coi.Spec.Windows.Devices)
+	}
+	if err := installWindowsDrivers(ctx, coi, r); err != nil {
+		return nil, err
+	}
+	deviceUtilPath, err := setupDeviceUtilTool(ctx, coi, r)
+	if err != nil {
+		return nil, err
+	}
+	return getChildrenDeviceLocationPaths(ctx, coi, vpciVMBusInstanceIDs, deviceUtilPath)
+}
+
+// getChildrenDeviceLocationPaths queries the UVM with the device-util tool with the formatted
+// parent bus device for the children devices' location paths from the uvm's view.
+// Returns a slice of WindowsDevices created from the resulting children location paths
+func getChildrenDeviceLocationPaths(ctx context.Context, coi *createOptionsInternal, vmBusInstanceIDs []string, deviceUtilPath string) ([]specs.WindowsDevice, error) {
+	p, l, err := createNamedPipeListener()
+	if err != nil {
+		return nil, err
+	}
+	defer l.Close()
+
+	var pipeResults []string
+	errChan := make(chan error)
+
+	go readCsPipeOutput(l, errChan, &pipeResults)
+
+	args := createDeviceUtilChildrenCommand(deviceUtilPath, vmBusInstanceIDs)
+	req := &shimdiag.ExecProcessRequest{
+		Args:   args,
+		Stdout: p,
+	}
+	exitCode, err := ExecInUvm(ctx, coi.HostingSystem, req)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to find devices with exit code %d", exitCode)
+	}
+
+	// wait to finish parsing stdout results
+	select {
+	case err := <-errChan:
+		if err != nil {
+			return nil, err
+		}
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+
+	// convert stdout results into windows devices
+	results := []specs.WindowsDevice{}
+	for _, value := range pipeResults {
+		specDev := specs.WindowsDevice{
+			ID:     value,
+			IDType: uvm.VPCILocationPathIDType,
+		}
+		results = append(results, specDev)
+	}
+	log.G(ctx).WithField("parsed devices", results).Info("found child assigned devices")
+	return results, nil
+}
+
+func createNamedPipeListener() (string, net.Listener, error) {
+	g, err := guid.NewV4()
+	if err != nil {
+		return "", nil, err
+	}
+	p := `\\.\pipe\` + g.String()
+	l, err := winio.ListenPipe(p, nil)
+	if err != nil {
+		return "", nil, err
+	}
+	return p, l, nil
+}
+
+// readCsPipeOutput is a helper function that connects to a listener and reads
+// the connection's comma separated outut until done. resulting comma separated
+// values are returned in the `result` param. The `done` param is used by this
+// func to indicate completion.
+func readCsPipeOutput(l net.Listener, errChan chan<- error, result *[]string) {
+	defer close(errChan)
+	c, err := l.Accept()
+	if err != nil {
+		errChan <- errors.Wrapf(err, "failed to accept named pipe")
+		return
+	}
+	r := bufio.NewReader(c)
+	var readErr error
+	var rawElem string
+	for {
+		rawElem, readErr = r.ReadString(',')
+		if len(rawElem) != 0 {
+			// remove the comma at the end of the line
+			elem := string(rawElem[:len(rawElem)-1])
+			*result = append(*result, elem)
+		}
+		if readErr != nil {
+			break
+		}
+	}
```

Comment:
You're going to hate me but I still think this can just be replaced with 
```go
b, err := ioutil.ReadAll(c)
if err != nil {
         errChan <- err
         return
}
res := strings.Split(string(b), ",")
// Add elems to *result
// VVV
```

Reply:
Updated